### PR TITLE
zn: implement Str <-> names conversinos

### PR DESCRIPTION
**Pull request by MaxVerevkin**
Plus deduplicate some code with macros.

### CODE REVIEW

**Comment by zeenix** [comment]
FYI not at all a problem (actually even better) I did **not** ask to split in a separate PR, only commits. :) Perhaps I should have spelled it out. :)

**Comment by MaxVerevkin** [comment]
Yeah, but I thought it's better to split the discussion in two "directions"

File: zbus_names/src/bus_name.rs [hunk 1]
```diff
@@ -213,49 +212,39 @@ impl<'s> TryFrom<&'s str> for BusName<'s> {
...
-                    Err(Error::InvalidBusName(unique_err, well_known_err))
-                }
-                Err(e) => Err(e),
-                Ok(name) => Ok(BusName::WellKnown(name)),
-            },
-            Err(e) => Err(e),
-            Ok(name) => Ok(BusName::Unique(name)),
-        }
+        let value: Str<'s> = value.into();
+        value.try_into()
```
Comment (zeenix) [inline zbus_names/src/bus_name.rs:238]: The BusName changes are not related to the de-duplicating ones, hence separate commits please. The "and" in the commit shortlog is typically a telltale sign of multiple logical changes.

File: zvariant/src/str.rs [hunk 2]
```diff
@@ -159,6 +160,15 @@ impl<'a> From<Arc<str>> for Str<'a> {
     }
 }
 
+impl<'a> From<Cow<'a, str>> for Str<'a> {
```
Comment (zeenix) [inline zvariant/src/str.rs:163]: this change is also a separate logical change, especially since it's a different crate.

**Review by zeenix — CHANGES_REQUESTED** [review]
Thanks so much for doing this work! Much appreciated but as usual, some nits. :)

File: zbus_names/src/unique_name.rs [hunk 3]
```diff
@@ -344,6 +344,18 @@ impl TryFrom<Arc<str>> for OwnedUniqueName {
     }
 }
 
+impl<'name> From<UniqueName<'name>> for Str<'name> {
+    fn from(value: UniqueName<'name>) -> Self {
+        value.0
+    }
+}
```
[the remark this answers is not in the document]
Comment (MaxVerevkin) [reply zbus_names/src/unique_name.rs]: I did, but it ended up in the third commit instead of first. Not sure how to resolve it.

[the remark this answers is not in the document]
Comment (MaxVerevkin) [reply zbus_names/src/unique_name.rs]: I hate it :sob:

image

**Comment by MaxVerevkin** [comment]
It's easier to do it from scratch than to unfuck this git history

File: zbus_names/src/unique_name.rs [hunk 4]
```diff
@@ -344,6 +344,18 @@ impl TryFrom<Arc<str>> for OwnedUniqueName {
...
+
+impl From<OwnedUniqueName> for Str<'static> {
+    fn from(value: OwnedUniqueName) -> Self {
+        value.0 .0
```
[the remark this answers is not in the document]
Comment (zeenix) [reply zbus_names/src/unique_name.rs:296]: > You mean `.0 .0`? 

Yeah.

> That't not me, that's rustfmt :)

Interesting. :) You probably want to use `into_inner()` or `into()` anyway. `.0.0` looks strange no matter how many spaces clippy adds into it. :laughing:

**Comment by zeenix** [comment]
> It's easier to do it from scratch than to unfuck this git history

Sometimes, yes. I sometimes just `git reset main` on my branch and re-commit everything using `git commit -c` (just have to keep the hashes around so you know which one to pass to `-c` option).

File: zbus_names/src/bus_name.rs [hunk 5]
```diff
@@ -216,63 +216,35 @@ impl<'s> TryFrom<&'s str> for BusName<'s> {
...
 
     fn try_from(value: Value<'s>) -> Result<Self> {
-        let value = Str::try_from(value)?;
-        Ok(match BusName::try_from(value.as_str())? {
-            BusName::Unique(_) => BusName::Unique(UniqueName::from_string_unchecked(value.into())),
-            BusName::WellKnown(_) => {
-                BusName::WellKnown(WellKnownName::from_string_unchecked(value.into()))
-            }
-        })
+        let value: Str<'s> = value.try_into()?;
```
Comment (zeenix) [inline zbus_names/src/bus_name.rs:246]: nit: I prefer:

```rust
let value = Str::try_from(value)?;
```

but here I think a combinator is in order:

```rust
Str::try_from(value).and_then(TryInto::try_into)
```

**Review by zeenix — CHANGES_REQUESTED** [review]
Looks perfect otherwise. 2 small nits and we're good to go. :)

File: zbus_names/src/bus_name.rs [hunk 6]
```diff
@@ -424,6 +424,14 @@ impl TryFrom<String> for OwnedBusName {
     }
 }
 
+impl TryFrom<Cow<'_, str>> for OwnedBusName {
+    type Error = Error;
+
+    fn try_from(value: Cow<'_, str>) -> Result<Self> {
+        Ok(Self::from(BusName::try_from(value)?))
```
Comment (zeenix) [inline zbus_names/src/bus_name.rs:414]: opportunity for combinators missed here too btw. ;) 

```rust
BusName::try_from(value).map(Self::from)
```

File: zbus_names/src/bus_name.rs [hunk 7]
```diff
@@ -194,7 +194,24 @@ impl<'name> From<WellKnownName<'name>> for BusName<'name> {
     }
 }
 
-/// Try to create an `BusName` from a string.
```
Comment (zeenix) [inline zbus_names/src/bus_name.rs:197]: why remove this comment? :thinking:

**Review by zeenix — CHANGES_REQUESTED** [review]
very close..

File: zbus_names/src/bus_name.rs (hunk 5 above)
Comment (zeenix) [reply zbus_names/src/bus_name.rs]: don't think this was resolved. :thinking:

Comment (MaxVerevkin) [reply zbus_names/src/bus_name.rs]: I thought this only applied to `try_into()`, not `into()`, will update.

File: zbus_names/src/bus_name.rs (hunk 7 above)
Comment (MaxVerevkin) [reply zbus_names/src/bus_name.rs:197]: For consistency - `TryFrom<String>`, `TryFrom<Arg<str>>`, etc. did not have this comment. And I didn't see much value in this comment. In general `///` comments for trait implementations aren't very useful as it should be clear what the trait does anyway.

Comment (zeenix) [reply zbus_names/src/bus_name.rs:197]: Right but please do this in a separate commit and add this comment as justification.

Comment (zeenix) [reply zbus_names/src/bus_name.rs:197]: Nice! Hate to bring it up but you forgot the break the long line in the commit log. Since rest of the commits break at 74, this will look very strange there.

**Review by zeenix — COMMENTED** [review]
Just update that one commit message and we're good to merge. Thanks for your patience.

**Review by zeenix — APPROVED** [review]
